spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,381 @@
+				Conversions.stringToBytes(value, this.encoding));
+		this.updateMap.put(key, new LocalChildData(value, stat.getVersion()));
+	}
+
+	private void createNode(String key, String value) throws Exception {
+		this.client.create().forPath(getPath(key), Conversions.stringToBytes(value, this.encoding));
+		this.updateMap.put(key, new LocalChildData(value, 0));
+	}
+
+	public String getPath(String key) {
+		return "".equals(key) ? this.root : this.root + "/" + key;
+	}
+
+
+	@Override
+	public boolean isAutoStartup() {
+		return autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		if (this.running) {
+			synchronized (this.lifecycleMonitor) {
+				if (this.running) {
+					if (this.cache != null) {
+						CloseableUtils.closeQuietly(this.cache);
+					}
+					this.cache = null;
+					this.running = false;
+				}
+			}
+		}
+	}
+
+	@Override
+	public void start() {
+		if (!this.running) {
+			synchronized (this.lifecycleMonitor) {
+				if (!this.running) {
+					try {
+						EnsurePath ensurePath = new EnsurePath(root);
+						ensurePath.ensure(client.getZookeeperClient());
+						this.cache = new PathChildrenCache(this.client, this.root, true);
+						this.cache.getListenable().addListener(new MetadataStoreListenerInvokingPathChildrenCacheListener());
+						this.cache.start(PathChildrenCache.StartMode.BUILD_INITIAL_CACHE);
+						running = true;
+					}
+					catch (Exception e) {
+						throw new ZookeeperMetadataStoreException("Exception while starting bean", e);
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		stop(null);
+	}
+
+	@Override
+	public boolean isRunning() {
+		return running;
+	}
+
+	@Override
+	public int getPhase() {
+		return phase;
+	}
+
+	private static class LocalChildData {
+
+		private String value;
+
+		private int version;
+
+		public LocalChildData(String value, int version) {
+			this.value = value;
+			this.version = version;
+		}
+
+		public String getValue() {
+			return this.value;
+		}
+
+		public int getVersion() {
+			return this.version;
+		}
+	}
+
+	private class MetadataStoreListenerInvokingPathChildrenCacheListener implements PathChildrenCacheListener {
+		@Override
+		public void childEvent(CuratorFramework client, PathChildrenCacheEvent event) throws Exception {
+			synchronized (ZookeeperMetadataStore.this.updateMap) {
+				switch (event.getType()) {
+					case CHILD_ADDED:
+						if (ZookeeperMetadataStore.this.updateMap.containsKey(getKey(event.getData().getPath()))) {
+							if (event.getData().getStat().getVersion() >=
+									ZookeeperMetadataStore.this.updateMap.get(getKey(event.getData().getPath())).getVersion()) {
+								ZookeeperMetadataStore.this.updateMap.remove(event.getData().getPath());
+							}
+						}
+						for (MetadataStoreListener listener : ZookeeperMetadataStore.this.listeners) {
+							listener.onAdd(getKey(event.getData().getPath()), Conversions.bytesToString(event.getData().getData(), encoding));
```

Comment:
We can optimize these lines a bit extracting `key` var for those `getKey(event.getData().getPath())` invocations.